File: types/zipline.ts
import type { UploadFileOptions } from "@/functions/zipline/files";
import type { Mimetypes } from "@/types/mimetypes";

export type DashURL = `http${"s" | ""}://${string}.${string}`;

export type APISelfUser = Omit<APIUser, "avatar">;

export interface APILoginSuccessResponse {
	user: APIUser;
	totp?: undefined;
}

export interface APILoginTotpResponse {
	totp: true;
	user?: undefined;
}

export type APILoginResponse = APILoginSuccessResponse | APILoginTotpResponse;

export interface APITokenResponse {
	token: string;
}

export interface APIUser {
	id: string;
	username: string;
	createdAt: string;
	updatedAt: string;
	role: "SUPERADMIN" | "ADMIN" | "USER";
	view: APIUserView;
	oauthProviders: OAuthProvider[];
	totpSecret: string | null;
	passkeys: Passkey[];
	quota: APIUserQuota | null;
	sessions: string[];
	avatar: string | null;
}

export interface APIUserQuota {
	id: string;
	createdAt: string;
	updatedAt: string;
	filesQuota: "BY_BYTES" | "BY_FILES";
	maxBytes: string | null;
	maxFiles: number | null;
	maxUrls: number | null;
	userId: "cm5ohf5r90003pl01r1vn6yus";
}

export interface APIUserView {
	enabled?: boolean;
	align?: "left" | "center" | "right";
	showMimetype?: boolean;
	content?: string;
	embed?: boolean;
	embedTitle?: string;
	embedDescription?: string;
	embedColor?: string;
	embedSiteName?: string;
}

export interface OAuthProvider {
	id: string;
	createdAt: string;
	updatedAt: string;
	userId: string;
	provider: string;
	username: string;
	accessToken: string;
	refreshToken: string;
	oauthId: string;
}

export interface Passkey {
	id: string;
	createdAt: string;
	updatedAt: string;
	lastUsed: string;
	name: string;
	reg: PasskeyReg;
	userId: string;
}

export interface PasskeyReg {
	id: string;
	type: string;
	rawId: string;
	response: PasskeyRegResponse;
	clientExtensionResults: PasskeyRegClientExtensionResults;
	authenticatorAttachment: string;
}

export interface PasskeyRegResponse {
	transports: string[];
	clientDataJSON: string;
	attestationObject: string;
}

export interface PasskeyRegClientExtensionResults {
	credProps: unknown | null;
}

export type APIRecentFiles = APIFile[];

export interface APIFile {
	createdAt: string;
	updatedAt: string;
	deletesAt: string | null;
	favorite: boolean;
	id: string;
	originalName: string | null;
	name: string;
	size: number;
	type: string;
	views: number;
	maxViews: number | null;
	folderId: string | null;
	thumbnail: string | null;
	tags: APITag[];
	password: boolean;
	url: string;
}

export interface APIUserStats {
	filesUploaded: number;
	favoriteFiles: number;
	views: number;
	avgViews: number;
	storageUsed: number;
	avgStorageUsed: number;
	urlsCreated: number;
	urlViews: number;
	sortTypeCount: {
		[key: string]: number;
	};
}

export interface APIUploadFile {
	id: APIFile["id"];
	type: keyof Mimetypes;
	url: string;
	pending: boolean;
}

export interface APIUploadResponse {
	files: APIUploadFile[];
	partialSuccess?: boolean;
	assumedMimetypes: boolean[];
	deletesAt?: string;
}

export type APIUploadPartialResponse = APIUploadResponse & {
	partialSuccess?: boolean;
	partialIdentifier?: string;
};

export interface APIShortenResponse {
	id: string;
	createdAt: string;
	updatedAt: string;
	code: string;
	vanity: string | null;
	destination: string;
	views: number;
	maxViews: number | null;
	enabled: true;
	userId: string;
	url: string;
}

export interface APITransactionResult {
	count: number;
	name?: string;
}

export interface APISettings {
	settings: {
		coreReturnHttpsUrls: boolean;
		coreDefaultDomain: string | null;
		coreTempDirectory: string;
		chunksEnabled: boolean;
		chunksMax: number | string;
		chunksSize: number | string;
		tasksDeleteInterval: number | string;
		tasksClearInvitesInterval: number | string;
		tasksMaxViewsInterval: number | string;
		tasksThumbnailsInterval: number | string;
		tasksMetricsInterval: number | string;
		filesRoute: string;
		filesLength: number;
		filesDefaultFormat: "random" | "uuid" | "date" | "name" | "gfycat";
		filesDisabledExtensions: string[]; // ???
		filesMaxFileSize: number | string;
		filesDefaultExpiration: string | null;
		filesAssumeMimetypes: boolean;
		filesDefaultDateFormat: string;
		filesRemoveGpsMetadata: boolean;
		filesRandomWordsNumAdjectives: number;
		filesRandomWordsSeparator: string;
		urlsRoute: string;
		urlsLength: number;
		featuresImageCompression: boolean;
		featuresRobotsTxt: boolean;
		featuresHealthcheck: boolean;
		featuresUserRegistration: boolean;
		featuresOauthRegistration: boolean;
		featuresDeleteOnMaxViews: boolean;
		featuresThumbnailsEnabled: boolean;
		featuresThumbnailsNumberThreads: number;
		featuresMetricsEnabled: boolean;
		featuresMetricsAdminOnly: boolean;
		featuresMetricsShowUserSpecific: boolean;
		featuresVersionChecking: boolean;
		featuresVersionAPI: string;
		invitesEnabled: boolean;
		invitesLength: number;
		websiteTitle: string;
		websiteTitleLogo: string | null;
		websiteExternalLinks: ExternalLink[];
		websiteLoginBackground: string | null;
		websiteLoginBackgroundBlur: boolean;
		websiteDefaultAvatar: string | null;
		websiteTos: string | null;
		websiteThemeDefault: string;
		websiteThemeDark: string;
		websiteThemeLight: string;
		oauthBypassLocalLogin: boolean;
		oauthLoginOnly: boolean;
		oauthDiscordClientId: string | null;
		oauthDiscordClientSecret: string | null;
		oauthDiscordRedirectUri: string | null;
		oauthDiscordAllowedIds: string[]; // +++
		oauthDiscordDeniedIds: string[]; // +++
		oauthGoogleClientId: string | null;
		oauthGoogleClientSecret: string | null;
		oauthGoogleRedirectUri: string | null;
		oauthGithubClientId: string | null;
		oauthGithubClientSecret: string | null;
		oauthGithubRedirectUri: string | null;
		oauthOidcClientId: string | null;
		oauthOidcClientSecret: string | null;
		oauthOidcAuthorizeUrl: string | null;
		oauthOidcTokenUrl: string | null;
		oauthOidcUserinfoUrl: string | null;
		oauthOidcRedirectUri: string | null;
		mfaTotpEnabled: boolean;
		mfaTotpIssuer: string;
		mfaPasskeys: boolean;
		ratelimitEnabled: boolean;
		ratelimitMax: number;
		ratelimitWindow: number | null;
		ratelimitAdminBypass: boolean;
		ratelimitAllowList: string[];
		httpWebhookOnUpload: string | null;
		httpWebhookOnShorten: string | null;
		discordWebhookUrl: string | null;
		discordUsername: string | null;
		discordAvatarUrl: string | null;
		discordOnUploadWebhookUrl: string | null;
		discordOnUploadUsername: string | null;
		discordOnUploadAvatarUrl: string | null;
		discordOnUploadContent: string | null;
		discordOnUploadEmbed: UploadEmbed | null; // check embed
		discordOnShortenWebhookUrl: string | null;
		discordOnShortenUsername: string | null;
		discordOnShortenAvatarUrl: string | null;
		discordOnShortenContent: string | null;
		discordOnShortenEmbed: ShortenEmbed | null; // check embed
		pwaEnabled: boolean;
		pwaTitle: string;
		pwaShortName: string;
		pwaDescription: string;
		pwaThemeColor: string;
		pwaBackgroundColor: string;
		domains: string[]; // +++
	};
	tampered: (keyof APISettings["settings"])[];
}

export interface ExternalLink {
	url: string;
	name: string;
}

export interface UploadEmbed {
	url: boolean;
	color: string | null;
	title: string | null;
	footer: string | null;
	thumbnail: boolean;
	timestamp: boolean;
	description: string | null;
	imageOrVideo: boolean;
}

export type ShortenEmbed = Omit<UploadEmbed, "imageOrVideo" | "thumbnail">;

export type APIUserNoIncl = Omit<
	APIUser,
	"view" | "oauthProviders" | "totpSecret" | "passkeys" | "sessions"
> & {
	passkeys: [];
	totpSecret: null;
	view: Record<string, never>;
	oauthProviders: [];
	sessions: [];
};

export type APIUsersNoIncl = APIUserNoIncl[];

export type APIUsers = APIUser[];

export interface APIURL {
	id: string;
	createdAt: string;
	updatedAt: string;
	code: string;
	vanity: string | null;
	destination: string;
	views: number;
	maxViews: number | null;
	password: string | null;
	enabled: boolean;
	userId: string;
}

export type APIURLs = APIURL[];

export interface APIFolder {
	id: string;
	createdAt: string;
	updatedAt: string;
	name: string;
	public: boolean;
	userId: string;
	files: APIFile[];
	allowUploads: boolean;
}

export type APIFolderNoIncl = Omit<APIFolder, "files">;

export type APIFolders = APIFolder[];

export type APIFoldersNoIncl = APIFolderNoIncl[];

export interface APITag {
	id: string;
	createdAt: string;
	updatedAt: string;
	name: string;
	color: string;
	files: TagFile[];
}

export type APITags = APITag[];

export interface TagFile {
	id: string;
}

export interface APIFiles {
	page: APIFile[];
	total: number;
	pages: number;
}

export interface APIStat {
	id: string;
	createdAt: string;
	updatedAt: string;
	data: APIStatData;
}

export type APIStats = APIStat[];

export interface APIStatData {
	users: number;
	files: number;
	fileViews: number;
	urls: number;
	urlViews: number;
	storage: number;
	filesUsers: APIStatUserFile[];
	urlsUsers: APIStatUserURL[];
	types: APIStatFileType[];
}

export interface APIStatUserFile {
	username: string;
	sum: number;
	storage: number;
	views: number;
}

export interface APIStatUserURL {
	username: string;
	sum: number;
	views: number;
}

export interface APIStatFileType {
	type: string;
	sum: number;
}

export interface APIInvite {
	id: string;
	createdAt: string;
	updatedAt: string;
	expiresAt: string | null;
	code: string;
	uses: number;
	maxUses: number;
	inviterId: string;
	inviter: APIInviteInvier;
}

export interface APIInviteInvier {
	username: string;
	id: string;
	role: APIUser["role"];
}

export type APIInvites = APIInvite[];

export interface APIExport {
	id: string;
	createdAt: string;
	updatedAt: string;
	completed: boolean;
	path: string;
	files: number;
	size: string;
	userId: string;
}

export type APIExports = APIExport[];

export interface ServerActionResponse {
	status: string;
}

export interface Preset {
	overrideDomain: UploadFileOptions["overrideDomain"];
	overrideFileName: UploadFileOptions["filename"];
	originalName: UploadFileOptions["originalName"];
	compression: UploadFileOptions["compression"];
	deletesAt: UploadFileOptions["expiresAt"];
	maxViews: UploadFileOptions["maxViews"];
	password: UploadFileOptions["password"];
	format: UploadFileOptions["format"];
	folder: UploadFileOptions["folder"];
	name: string;
}

export type APIVersion = OldV4APIVersion | NewV4APIVersion;

export interface OldV4APIVersion {
	version: string;
}

export interface NewV4APIVersion {
	data: APIVersionData;
	details: APIVersionDetails;
}

interface APIVersionDetails {
	version: string;
	sha: string;
}

interface APIVersionData {
	latest: APIVersionDataLatest;
	isUpstream: boolean;
	isRelease: boolean;
	isLatest: false;
	version: APIVersionDataVersion;
}

interface APIVersionDataLatest {
	tag: string;
	url: string;
	commit: APIVersionDataLatestCommit;
}

interface APIVersionDataLatestCommit {
	sha: string;
	url: string;
	pull: boolean;
}

interface APIVersionDataVersion {
	tag: string;
	sha: string;
	url: string;
}

export interface APIPublicSettings {
	oauth: {
		bypassLocalLogin: boolean;
		loginOnly: boolean;
	};
	oauthEnabled: {
		discord: boolean;
		github: boolean;
		google: boolean;
		oidc: boolean;
	};
	website: {
		loginBackground?: string | null;
		loginBackgroundBlur?: boolean;
		title?: string;
		tos: boolean;
	};
	features: {
		oauthRegistration: boolean;
		userRegistration: boolean;
	};
	mfa: {
		passkeys: boolean;
	};
	tos?: string | null;
	files: {
		maxFileSize: string;
		defaultFormat:
			| "uuid"
			| "date"
			| "random"
			| "name"
			| "gfycat"
			| "random-words";
	};
	chunks: {
		max?: string;
		size?: string;
		enabled?: boolean;
	};
	domains?: string[];
	firstSetup: boolean;
}

export interface APIWebSettings {
	config: {
		chunks: {
			max: string;
			size: string;
			enabled: boolean;
		};
		tasks: {
			deleteInterval: string;
			clearInvitesInterval: string;
			maxViewsInterval: string;
			thumbnailsInterval: string;
			metricsInterval: string;
		};
		files: {
			route: string;
			length: number;
			defaultFormat:
				| "uuid"
				| "date"
				| "random"
				| "name"
				| "gfycat"
				| "random-words";
			disabledExtensions: string[];
			maxFileSize: string;
			defaultExpiration: string | null;
			assumeMimetypes: boolean;
			defaultDateFormat: string;
			removeGpsMetadata: boolean;
			randomWordsNumAdjectives: number;
			randomWordsSeparator: string;
		};
		urls: {
			route: string;
			length: number;
		};
		features: {
			imageCompression: boolean;
			robotsTxt: boolean;
			healthcheck: boolean;
			userRegistration: boolean;
			oauthRegistration: boolean;
			deleteOnMaxViews: boolean;
			thumbnails: {
				enabled: boolean;
				num_threads: number;
			};
			metrics: {
				enabled: boolean;
				adminOnly: boolean;
				showUserSpecific: boolean;
			};
			versionChecking: boolean;
			versionAPI: string;
		};
		invites: {
			enabled: boolean;
			length: number;
		};
		website: {
			title: string;
			titleLogo: string | null;
			externalLinks: {
				name: string;
				url: string;
			}[];
			loginBackground: string | null;
			loginBackgroundBlur: boolean;
			defaultAvatar: string | null;
			theme: {
				default: string;
				dark: string;
				light: string;
			};
			tos: string | null;
		};
		mfa: {
			totp: {
				enabled: boolean;
				issuer: string;
			};
			passkeys: boolean;
		};
		pwa: {
			enabled: boolean;
			title: string;
			shortName: string;
			description: string;
			themeColor: string;
			backgroundColor: string;
		};
		oauthEnabled: {
			discord: boolean;
			github: boolean;
			google: boolean;
			oidc: boolean;
		};
		oauth: {
			bypassLocalLogin: boolean;
			loginOnly: boolean;
		};
		version: string;
	};
	codeMap: {
		ext: string;
		mime: string;
		name: string;
	}[];
}

File: app/src/main/java/com/stefdp/zipline/utils/Zipline.kt
package com.stefdp.zipline.utils

import android.os.Parcelable
import com.stefdp.zipline.network.models.BaseFolder
import com.stefdp.zipline.network.models.FilesFormat
import com.stefdp.zipline.network.models.ThumbnailFormat
import com.stefdp.zipline.network.models.requests.UploadCompressionType
import io.github.z4kn4fein.semver.toVersion
import kotlinx.parcelize.Parcelize

@Parcelize
data class ZiplineViewState(
    val adminUsers: ZiplineViewStateType,
    val adminInvites: ZiplineViewStateType,
    val files: ZiplineViewStateType,
    val folders: ZiplineViewStateType,
    val urls: ZiplineViewStateType,
) : Parcelable

@Parcelize
enum class ZiplineViewStateType : Parcelable {
    COMPACT,
    LARGE;

    companion object {
        private val names = entries.map { it.name }.toSet()

        operator fun contains(value: String): Boolean = names.contains(value)
    }
}

val minimumZiplineVersion = "4.5.0".toVersion()

val inviteExpiresAtDates = listOf(
    "never" to "Never",
    "30min" to "30 minutes",
    "1h" to "1 hour",
    "6h" to "6 hours",
    "12h" to "12 hours",
    "1d" to "1 day",
    "3d" to "3 days",
    "5d" to "5 days",
    "7d" to "7 days",
)

val deletesAtDates = listOf(
    "never" to "Never",
    "5min" to "5 minutes",
    "10min" to "10 minutes",
    "15min" to "15 minutes",
    "30min" to "30 minutes",
    "1h" to "1 hour",
    "2h" to "2 hours",
    "3h" to "3 hours",
    "4h" to "4 hours",
    "5h" to "5 hours",
    "6h" to "6 hours",
    "8h" to "8 hours",
    "12h" to "12 hours",
    "1d" to "1 day",
    "3d" to "3 days",
    "5d" to "5 days",
    "7d" to "7 days",
    "1w" to "1 week",
    "1.5w" to "1.5 weeks",
    "2w" to "2 weeks",
    "3w" to "3 weeks",
    "30d" to "1 month (30 days)",
    "45.625d" to "1.5 months (~45 days)",
    "60d" to "2 months (60 days)",
    "90d" to "3 months (90 days)",
    "120d" to "4 months (120 days)",
    "0.5 year" to "6 months (0.5 year)",
    "1y" to "1 year"
)

val nameFormats = listOf(
    FilesFormat.RANDOM to "Random",
    FilesFormat.DATE to "Date",
    FilesFormat.UUID to "UUID",
    FilesFormat.NAME to "Use file name",
    FilesFormat.GFYCAT to "Gfycat-style name"
)

val compressionFormats = listOf(
    UploadCompressionType.JPG to ".jpg",
    UploadCompressionType.PNG to ".png",
    UploadCompressionType.WEBP to ".webp",
    UploadCompressionType.JXL to ".jxl"
)

val thumbnailFormats = listOf(
    ThumbnailFormat.JPG to ".jpg",
    ThumbnailFormat.PNG to ".png",
    ThumbnailFormat.WEBP to ".webp",
)

fun getFolderPath(
    folder: BaseFolder,
    folders: List<BaseFolder> = emptyList()
): String {
    if (folder.parentId == null) {
        return folder.name
    }

    val parent = folders.find { it.id == folder.parentId }

    return if (parent != null) {
        "${getFolderPath(parent, folders)} / ${folder.name}"
    } else {
        folder.name
    }
}

fun isChildOf(
    folder: BaseFolder,
    folders: List<BaseFolder>,
    targetParentId: String
): Boolean {
    if (folder.id == targetParentId) return true

    val parentId = folder.parentId ?: return false

    val parent = folders.find { it.id == parentId }

    return if (parent != null) {
        if (parent.id == targetParentId) true
        else isChildOf(parent, folders, targetParentId)
    } else {
        false
    }
}

val settingNames = listOf(
    "coreDefaultDomain" to "Default Domain",
    "coreReturnHttpsUrls" to "Return HTTPS URLs",
    "coreTempDirectory" to "Temporary Directory",

    "chunksEnabled" to "Enable Chunks",
    "chunksMax" to "Max Chunks Size",
    "chunksSize" to "Chunks Size",

    "tasksDeleteInterval" to "Delete Files Interval",
    "tasksClearInvitesInterval" to "Clear Invites Interval",
    "tasksMaxViewsInterval" to "Max Views Interval",
    "tasksThumbnailsInterval" to "Thumbnail Interval",
    "tasksMetricsInterval" to "Metrics Interval",

    "mfaPasskeys" to "Passkeys",
    "mfaTotpEnabled" to "Enable TOTP",
    "mfaTotpIssuer" to "Issuer",

    "featuresImageCompression" to "Image Compression",
    "featuresRobotsTxt" to "/robots.txt",
    "featuresHealthcheck" to "Healthcheck",
    "featuresUserRegistration" to "User Registration",
    "featuresOauthRegistration" to "OAuth Registration",
    "featuresDeleteOnMaxViews" to "Delete on Max Views",
    "featuresMetricsEnabled" to "Enable Metrics",
    "featuresMetricsAdminOnly" to "Admin Only Metrics",
    "featuresMetricsShowUserSpecific" to "Show User Specific Metrics",
    "featuresThumbnailsEnabled" to "Enable Thumbnails",
    "featuresThumbnailsNumberThreads" to "Thumbnails Number Threads",
    "featuresThumbnailsFormat" to "Thumbnails Format",
    "featuresVersionChecking" to "Version Checking",
    "featuresVersionAPI" to "Version API URL",

    "filesRoute" to "Route",
    "filesLength" to "Length",
    "filesAssumeMimetypes" to "Assume Mimetypes",
    "filesRemoveGpsMetadata" to "Remove GPS Metadata",
    "filesDefaultFormat" to "Default Format",
    "filesDisabledExtensions" to "Disabled Extensions",
    "filesMaxFileSize" to "Max File Size",
    "filesDefaultExpiration" to "Default Expiration",
    "filesDefaultDateFormat" to "Default Date Format",

    "urlsRoute" to "Route",
    "urlsLength" to "Length",

    "invitesEnabled" to "Enable Invites",
    "invitesLength" to "Length",

    "ratelimitEnabled" to "Enable Ratelimit",
    "ratelimitAdminBypass" to "Admin Bypass",
    "ratelimitMax" to "Max Requests",
    "ratelimitWindow" to "Window",
    "ratelimitAllowList" to "Allow List",

    "websiteTitle" to "Title",
    "websiteTitleLogo" to "Title Logo",
    "websiteExternalLinks" to "External Links",
    "websiteLoginBackground" to "Login Background",
    "websiteLoginBackgroundBlur" to "Login Background Blur",
    "websiteDefaultAvatar" to "Default Avatar",
    "websiteTos" to "Terms of Service",
    "websiteThemeDefault" to "Default Theme",
    "websiteThemeDark" to "Dark Theme",
    "websiteThemeLight" to "Light Theme",

    "oauthBypassLocalLogin" to "Bypass Local Login",
    "oauthLoginOnly" to "Login Only",

    "oauthDiscordClientId" to "Discord Client ID",
    "oauthDiscordClientSecret" to "Discord Client Secret",
    "oauthDiscordRedirectUri" to "Discord Redirect URL",

    "oauthGoogleClientId" to "Google Client ID",
    "oauthGoogleClientSecret" to "Google Client Secret",
    "oauthGoogleRedirectUri" to "Google Redirect URL",

    "oauthGithubClientId" to "GitHub Client ID",
    "oauthGithubClientSecret" to "GitHub Client Secret",
    "oauthGithubRedirectUri" to "GitHub Redirect URL",

    "oauthOidcClientId" to "OIDC Client ID",
    "oauthOidcClientSecret" to "OIDC Client Secret",
    "oauthOidcAuthorizeUrl" to "OIDC Authorize URL",
    "oauthOidcTokenUrl" to "OIDC Token URL",
    "oauthOidcUserinfoUrl" to "OIDC Userinfo URL",
    "oauthOidcRedirectUri" to "OIDC Redirect URL",

    "pwaEnabled" to "PWA Enabled",
    "pwaTitle" to "Title",
    "pwaShortName" to "Short Name",
    "pwaDescription" to "Description",
    "pwaThemeColor" to "Theme Color",
    "pwaBackgroundColor" to "Background Color",

    "httpWebhookOnUpload" to "On Upload",
    "httpWebhookOnShorten" to "On Shorten",

    "discordWebhookUrl" to "Webhook URL",
    "discordUsername" to "Username",
    "discordAvatarUrl" to "Avatar URL",

    "discordOnUploadWebhookUrl" to "Webhook URL",
    "discordOnUploadAvatarUrl" to "Avatar URL",
    "discordOnUploadUsername" to "Username",
    "discordOnUploadContent" to "Content",
    "discordOnUploadEmbed" to "Embed",

    "discordOnUploadEmbed.color" to "Color",
    "discordOnUploadEmbed.description" to "Description",
    "discordOnUploadEmbed.footer" to "Footer",
    "discordOnUploadEmbed.imageOrVideo" to "Image/Video",
    "discordOnUploadEmbed.thumbnail" to "Thumbnail",
    "discordOnUploadEmbed.timestamp" to "Timestamp",
    "discordOnUploadEmbed.title" to "Title",
    "discordOnUploadEmbed.url" to "URL",

    "discordOnShortenWebhookUrl" to "Webhook URL",
    "discordOnShortenUsername" to "Username",
    "discordOnShortenAvatarUrl" to "Avatar URL",
    "discordOnShortenContent" to "Content",
    "discordOnShortenEmbed" to "Embed",

    "discordOnShortenEmbed.color" to "Color",
    "discordOnShortenEmbed.description" to "Description",
    "discordOnShortenEmbed.footer" to "Footer",
    "discordOnShortenEmbed.timestamp" to "Timestamp",
    "discordOnShortenEmbed.title" to "Title",
    "discordOnShortenEmbed.url" to "URL",
)

fun getSettingName(path: List<String>): String {
    val key = path.joinToString(".")

    return settingNames.find { it.first == key }?.second ?: key
}